packages/lib/src/computedTree/computedTree.ts

```diff
@@ -0,0 +1,57 @@
+import { computed } from "mobx"
+import { readonlyMiddleware } from "../actionMiddlewares/readonlyMiddleware"
+import { createContext } from "../context/context"
+import { isDataModelClass } from "../dataModel/utils"
+import { isModelClass } from "../model/utils"
+import { isTreeNode } from "../tweaker/core"
+import { tweak } from "../tweaker/tweak"
+import { failure } from "../utils"
+
+const computedTreeContext = createContext(false)
+
+/**
+ * Returns if a given node is a computed tree node.
+ *
+ * @param node Node to check.
+ * @returns `true` if it is a computed tree node, `false` otherwise.
+ */
+export function isComputedTreeNode(node: object): boolean {
+  return computedTreeContext.get(node)
+}
+
+/**
+ * Decorator for turning a computed property into a computed tree which supports tree traversal
+ * functions, contexts, references, etc.
+ *
+ * @param target Prototype of the class.
+ * @param propertyKey Name of the member.
+ * @param descriptor Property descriptor for the member.
+ */
+export function computedTree(
+  target: any,
+  propertyKey: string,
+  descriptor: PropertyDescriptor
+): void {
```

Comment:
are you sure this decorator works with both babel and tsc? (I think they used to generate different descriptor code)

Reply:
Good point, I hadn't considered that. I'll check and get back to you.

Reply:
I've checked how decorators are implemented by Babel, mainly by reverse-engineering it via the Babel REPL. TS + decorators (legacy) + loose (`true` & `false`) has the same decorator signature as [TS accessor decorators](https://www.typescriptlang.org/docs/handbook/decorators.html#accessor-decorators) (see [REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.6&spec=false&loose=true&code_lz=GYVwdgxgLglg9mABAEwKYTgJwIZSwCgEpEBvAKEUQzAGc4AbVAOnrgHN9tM2QBbVMFBqEyAXzJkI9bDRqIAgqQqIAAmgw48mZW1RREANyJLKASEx6QmJAFZl48ZIR1GLdvjCoA7gqJMDhEA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=env%2Creact%2Cstage-0%2Cstage-1%2Cstage-2%2Ctypescript&prettier=true&targets=&version=7.17.1&externalPlugins=&assumptions=%7B%7D)). So I'd say the current implementation should work for both TSC and Babel, but I don't have much experience with the quirks of decorators yet.

Do you want to support other decorator proposals than "legacy"?

It would be nice to test this as part of CI/CD though. :thinking:

Reply:
Those should be good enough for now :)